### PR TITLE
Spark: Support ORC vectorized reads

### PR DESCRIPTION
**Pull request by shardulm94**
- Most of the new code is added in `VectorizedSparkOrcReaders`
- Replaced `NullValuesColumnVector` with `ConstantColumnVector` which can support any constant (including nulls)
- Moved the Iceberg to Spark constant conversion logic from `RowDataReader` into `BaseDataReader` so that it can be reused in `BatchDataReader`
- Modified many of the Spark read test cases to test for both vectorized and non vectorized codepaths
- Modified `TestFilteredScan` in Spark3 to use IcebergGenerics instead of Avro just like in Spark2. This enables use to test ORC reads/writes which do not have Avro GenericRecord writer.

Benchmark results:
Tests reads of 10 files with 5M records each
```
Benchmark                                                                          Mode  Cnt    Score    Error  Units
IcebergSourceFlatORCDataReadBenchmark.readFileSourceNonVectorized                    ss    5   42.789 ±  3.294   s/op
IcebergSourceFlatORCDataReadBenchmark.readFileSourceVectorized                       ss    5   18.566 ±  1.450   s/op
IcebergSourceFlatORCDataReadBenchmark.readIcebergNonVectorized                       ss    5   30.186 ±  1.007   s/op
IcebergSourceFlatORCDataReadBenchmark.readIcebergVectorized                          ss    5   18.835 ±  0.818   s/op
IcebergSourceFlatORCDataReadBenchmark.readWithProjectionFileSourceNonVectorized      ss    5    8.935 ±  0.801   s/op
IcebergSourceFlatORCDataReadBenchmark.readWithProjectionFileSourceVectorized         ss    5    2.387 ±  0.195   s/op
IcebergSourceFlatORCDataReadBenchmark.readWithProjectionIcebergNonVectorized         ss    5   10.691 ±  0.603   s/op
IcebergSourceFlatORCDataReadBenchmark.readWithProjectionIcebergVectorized            ss    5    2.653 ±  0.511   s/op
IcebergSourceNestedORCDataReadBenchmark.readFileSourceNonVectorized                  ss    5  118.318 ±  1.583   s/op
IcebergSourceNestedORCDataReadBenchmark.readFileSourceVectorized                     ss    5   98.858 ± 12.668   s/op
IcebergSourceNestedORCDataReadBenchmark.readIcebergNonVectorized                     ss    5   18.943 ±  1.305   s/op
IcebergSourceNestedORCDataReadBenchmark.readIcebergVectorized                        ss    5    9.330 ±  0.938   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceNonVectorized    ss    5   86.136 ±  1.139   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceVectorized       ss    5   81.307 ± 14.090   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionIcebergNonVectorized       ss    5   16.671 ±  0.855   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionIcebergVectorized          ss    5    6.710 ±  0.679   s/op
```

### CODE REVIEW

**Comment by rdblue** [comment]
Thanks, @shardulm94! It's great to see a PR for this. I'll try to get some time to review it.

**Comment by rdblue** [comment]
Do we know why the nested data benchmarks show such a big improvement? Are we running correctness tests for the same cases to make sure we aren't dropping data by accident? It just seems a bit too good.

**Comment by shardulm94** [comment]
Which cases are you comparing for nested data? `readIcebergNonVectorized` v/s `readIcebergVectorized` for nested data shows 2-3x improvement which is similar to the improvements for flat data. For nested data the `readFileSourceVectorized` and `readWithProjectionFileSourceVectorized` are not really relevant since the file source defaults to row by row reading for nested data, so I guess we should just remove them. I modified the Spark unit tests to also test the vectorized codepaths, so I am assuming those tests check correctness, but I can do some sanity checks manually.

**Comment by rdblue** [comment]
> For nested data the readFileSourceVectorized and readWithProjectionFileSourceVectorized are not really relevant since the file source defaults to row by row reading for nested data

That's the answer I was looking for. It was really surprising that Spark vectorized was so slow. Thanks!

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java [hunk 1]
```diff
@@ -177,9 +179,20 @@ public ReadBuilder filter(Expression newFilter) {
       return this;
     }
 
+    public ReadBuilder createBatchedReaderFunc(Function<TypeDescription, OrcBatchReader<?>> batchReaderFunction) {
+      this.batchedReaderFunc = batchReaderFunction;
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/ORC.java:183]: This should validate that `createReaderFunc` hasn't also been called. And that function should validate that this one hasn't been called. That way there is no ambiguous behavior and the user gets an error when they try to set up both batch and row readers.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java [hunk 2]
```diff
@@ -42,19 +42,26 @@
 import org.apache.spark.sql.types.Decimal;
 import org.apache.spark.unsafe.types.UTF8String;
 
-
-class SparkOrcValueReaders {
+public class SparkOrcValueReaders {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:45]: I like the changes here and in `SparkOrcReader`, but they aren't really needed for this PR and increase the number of files that are touched. In general, I prefer keeping changes like this separate to avoid unnecessary conflicts, and so that we can revert commits without removing unrelated clean-up.

File: orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java [hunk 3]
```diff
@@ -130,4 +140,25 @@ public T next() {
     }
   }
 
+  private static class OrcBatchIterator<T> implements Iterator<T> {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java:143]: This needs to be a `CloseableIterator` to avoid leaking an open file. I'm also wondering if we could replace this implementation with a method like `CloseableIterable.transform` for `CloseableIterator`. All this is doing is calling a method on the result of another `Iterator`.

File: spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+import org.apache.spark.sql.vectorized.ColumnarMap;
+import org.apache.spark.unsafe.types.UTF8String;
+
+public class VectorizedSparkOrcReaders {
+
+  private VectorizedSparkOrcReaders() {
+  }
+
+  public static OrcBatchReader<ColumnarBatch> buildReader(Schema expectedSchema, TypeDescription fileSchema,
+      Map<Integer, ?> idToConstant) {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:52]: Nit: we would normally align this with the start of `Schema` on the previous line.

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java [hunk 5]
```diff
@@ -75,6 +103,16 @@
       }
 
       iter = builder.build();
+    } else if (task.file().format() == FileFormat.ORC) {
+      iter = ORC.read(location)
+          .project(expectedSchema)
```
Comment (shardulm94) [inline spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:108]: I will change this to not project constant columns to avoid materializing a `ColumnVector` similar to what we did in #1191

File: spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    @Override
+    public boolean getBoolean(int rowId) {
+      Boolean value = (Boolean) primitiveValueReader.read(vector, rowId);
+      return value != null ? value : false;
+    }
+
+    @Override
+    public int getInt(int rowId) {
+      Integer value = (Integer) primitiveValueReader.read(vector, rowId);
+      return value != null ? value : 0;
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:282]: Why this extra check just to add a default value? If it is okay to return a default value, then `getInt` should never be called for a `rowId` where the value is null. And if this is never called when the value is null, then I'd rather directly cast. That way, a `NullPointerException` is thrown if the method contract is violated instead of silently returning the wrong value.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java (hunk 2 above)
Comment (shardulm94) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:45]: I wanted to avoid duplicating the `if (precision < 18) then Decimal18Reader else Decimal38Reader` logic, since I am reusing it in `VectorizedSparkOrcReaders`. Similarly I made `utf8string` and `timestampTzs` methods public for reuse in the `vectorized` package. Let me know if you would like me to revert the Decimal change.

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:45]: Okay, I think that's a good justification for including this. Thanks!

**Comment by rdblue** [comment]
The only blocker I see is that the batch iterator doesn't pass the close call down to the underlying ORC reader. Otherwise, everything is minor.

Awesome work, @shardulm94!

File: orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java (hunk 3 above)
Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java:143]: This `Iterator` is converted to a `CloseableIterator` at https://github.com/apache/iceberg/blob/6a5e8c53d3950922b8f6bd753329b7d943ac423d/orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java#L89
I too want to avoid creating the `OrcBatchIterator` class, however I am unable to make `Iterators.transform` work here because of the capture type of `batchReaderFunction`.

image

I am also not able to cast the `batchReaderFunction` to the generic type used by `OrcIterable`.

Comment (rdblue) [reply orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java:143]: What about `Iterators.transform(rowBatchIterator, batchReaderFunction.apply(readOrcSchema)::read)`? That should allow the compiler to convert it to a Function.

Comment (rdblue) [reply orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java:143]: > This Iterator is converted to a CloseableIterator at

Yes, it is converted to a `CloseableIterator`, but that iterator's close method won't do anything unless this is `Closeable`. Here's where it delegates:

```java
      @Override
      public void close() throws IOException {
        if (iterator instanceof Closeable) {
          ((Closeable) iterator).close();
        }
      }
```

So if this `Iterator` doesn't implement `Close` to pass the close to `batchIter`, then `batchIter` won't get closed.

Alternatively, you could add a `transform` factory method to `CloseableIterator`:

```java
  static <S, T> CloseableIterator<T> transform(CloseableIterator<S> iter, Function<S, T> transform) {
    return new CloseableIterator<T>() {
      @Override
      public void close() throws IOException {
        iter.close();
      }

      @Override
      public boolean hasNext() {
        return iter.hasNext();
      }

      @Override
      public T next() {
        return transform.apply(iter.next());
      }
    };
  }
```

Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java:143]: I remember trying this too. `OrcBatchReader` is a functional interface, so it should not have issues converting the reader to a Function. The issue seems to be the capture type.

image

Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java:143]: Okay, I am able to make it work if I assign the `transformed` iterator to `Iterator` instead of `Iterator<T>`.

File: spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java (hunk 6 above)
Comment (shardulm94) [reply spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:282]: Yep! That makes sense. If the reader is not going to check for null, then returning a default value would be erroneous. One issue I see is that Spark's own ColumnVector code seems to violate this at https://github.com/apache/spark/blob/d5b903e38556ee3e8e1eb8f71a08e232afa4e36a/sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java#L92 and similar methods within the same class.
In Spark 2.4 these methods are not actually used, the methods are used at https://github.com/apache/spark/blob/d5b903e38556ee3e8e1eb8f71a08e232afa4e36a/sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnarArray.java#L53 but those are not used anywhere.

In Spark 3.0, they are used at https://github.com/apache/spark/blob/3b0aee3f9557ae1666b63665b08d899fe7682852/sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnarArray.java#L55, but this seems incorrect for the reasons you mentioned above. I am unsure when the `copy` method would actually be triggered though.

File: spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java (hunk 4 above)
Comment (shardulm94) [reply spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java]: Fixed in https://github.com/apache/iceberg/pull/1189/commits/faeffa368d4bc7766ea2008b351acddd792ad7ef

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java (hunk 1 above)
Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/ORC.java:187]: Fixed in https://github.com/apache/iceberg/pull/1189/commits/d13c3662ac5d7417c647ad5c6c5b22685341dc46

File: spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java (hunk 5 above)
Comment (shardulm94) [reply spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java]: Fixed in https://github.com/apache/iceberg/pull/1189/commits/9c2af8d5038b8dd478c39f57729370f7244816a7

File: orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java (hunk 3 above)
Comment (shardulm94) [reply orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java]: Fixed in https://github.com/apache/iceberg/pull/1189/commits/7d0cf1cafec9ff42d8e4cb5019185af00ff80d58

Added `CloseableIterator.transform`, updated `VectorizedRowBatchIterator` to use `CloseableIterator`, and also fixed the issue with incorrect use of `CloseableIterator.withClose`.

File: spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java (hunk 6 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:282]: Looks like other implementations in Spark just return from the underlying vector, without checking. So these methods would always return something. Arrow is an exception: if null checking is enabled (the default) then it will check and throw `IllegalStateException`.

I think it still makes sense to throw the `NullPointerException`, even if `copy` in Spark 3 would use it. I don't see any uses of `ArrayData.copy` in Spark 3, and any problem would only affect arrays, so the impact is limited. Plus, other sources (Arrow) break in this case and it appears to be overlooked. I think it makes sense to throw an exception to prevent a copy that corrupts the data by dropping null values.

**Comment by rdblue** [comment]
Nice work, @shardulm94! I'm merging this because the blockers have been resolved and this is large. We can follow up with a change for the way nulls are handled in getters. I don't think that is a blocker because there are examples in Spark of both ignoring the issue and throwing exceptions.

File: orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java (hunk 3 above)
Comment (megshao) [reply orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java]: PaymentFlowInfoObject 那邊～
